### DISCUSSION TITLE
query and zustand at the same time?

### DISCUSSION THREAD

**Discussion by johnf** [pmndrs/jotai#664] [Q&A]
Hi,

I've been using jotai and the query integration to cache data from my api. In this particular instance a user atom.
I now need to access that from outside react (push notification functions which needs to live outside of the render tree).

Is there a way to have connect an atomwithQuery to say zustand so it can be accessed from outside react?

**Comment by dai-shi** [comment]
Thanks for opening a discussion.

This sounds like a field where we still look for a best practice. Feedback and suggestions are welcome from anyone.

One idea would be to use `useEffect`.

```js
const fooAtom = atomWithQuery(...)

let globalToken

const Component = () => {
  const [foo] = useAtom(fooAtom)
  const { token } = foo
  useEffect(() => {
    globalToken = token
    return () => {
      globalToken = null
    }
  }, [token])
  ...
}
```

**Reply by johnf** [reply]
Thanks @dai-shi that seems really obvious in retrospect :)

This get's me most of the way there.
Will passing the dispatch to a global for the refetch also work? Or does that need to run inside the React context?

**Reply by dai-shi** [reply]
(a) As long as it's only one instance using the effect in JS memory, and (b) as long as the effect cleans up correctly, it should work without any issues. (a) and (b) are something only app developers can make sure.

**Comment by aulneau** [comment]
Outside of what @dai-shi suggested, another option to access query data outside of react would be to define a custom `QueryClient` and pass that to zustand. I have an example here (without the zustand bits) -> https://codesandbox.io/s/zealous-shaw-bsfv8?file=/src/App.tsx, lmk if this works for you!

**Reply by dai-shi** [reply]
I guess this is preferable in jotai/query use case.
